### PR TITLE
scripts/pythondistdeps: Preparing PR for upstream

### PR DESCRIPTION
**Pull request by torsava**
First two commits are upstreaming of changes from Fedora.

### CODE REVIEW

**Comment by torsava** [comment]
@hroncok if you want to take a look at what I have right now. I'm going to sleep :)

**Comment by hroncok** [comment]
Note to the future: Maintaining the test data and tests in upstream RPM repository and/or in downstream dist-git is going to be painful. For now, that's good, but in the future this needs to be solved somehow.

**Comment by torsava** [comment]
TBD: Move tests to some permanent location, ideally hook them into the existing test structure (if possible).

**Comment by torsava** [comment]
> Note to the future: Maintaining the test data and tests in upstream RPM repository and/or in downstream dist-git is going to be painful. For now, that's good, but in the future this needs to be solved somehow.

The tests I think are fine, do you see any concrete issues with having them alongside the code in the git repo? The test data I agree, dunno how to approach it better right now.

File: scripts/pythondistdeps.py
Comment (torsava) [inline scripts/pythondistdeps.py]: @hroncok What do you think?

Comment (hroncok) [reply scripts/pythondistdeps.py]: The content looks reasonable but I wouldn't put such a long comment to the actual code. Not sure where to put it, but I'd simply comment this to the final PR, explaining why we have not ported it to importlib metadata. WDYT?

Comment (torsava) [reply scripts/pythondistdeps.py]: I think it should be somewhere discoverable for when somebody else has the same idea 2 years from now. I agree the current location is not great.

How aobut adding a comment to the code: "# Notes from an attempted rewrite from pkg_resources to importlib.metadata in 2020 can be found in the commit message that added this code line"? and putting it in the commit message?

Comment (hroncok) [reply scripts/pythondistdeps.py]: Add a short note to pkg_reosurces import saying that a port to importlib.resources was attempted in 2020 and failed, see commit message for details ?

Comment (torsava) [reply scripts/pythondistdeps.py]: Exactly. Ok, doing it.

File: scripts/pythondistdeps.py
Comment (torsava) [inline scripts/pythondistdeps.py:17]: @hroncok Changed to this.

File: scripts/pythondistdeps.py
Comment (torsava) [inline scripts/pythondistdeps.py]: @hroncok And this.

**Comment by hroncok** [comment]
See also `https://github.com/rpm-software-management/rpm/pull/1015`.

**Comment by torsava** [comment]
> See also `https://github.com/rpm-software-management/rpm/pull/1015`.

Dayum, that's exactly what I was about to dig deep into. Well, good then :D I'll try to run it against the tests I prepared.

**Comment by hroncok** [comment]
> exactly what I was about to dig deep into

Glad I have saved the day! Was scared it will come too late.

**Comment by torsava** [comment]
> Glad I have saved the day! Was scared it will come too late.

I did some preliminary work, but nothing too big. I quickly realized I would have to drastically rewrite it like the gentleman in the PR, and decided to do the importlib first :) Thanks for alerting me!

**Comment by torsava** [comment]
Added the commit from this PR to properly test it: https://github.com/rpm-software-management/rpm/pull/1015

Added tests for version handling.

**Comment by torsava** [comment]
Note that currently the metadata for testing are saved in git. The plan is to remove them, and create a pytest fixture to download the metadata from PyPI before running the tests. Except for the `pyreq2rpm.tests` metadata which I've concocted myself.

**Comment by hroncok** [comment]
@torsava You've asked for some code similar to the planed pytest fixture. I only know about this: https://github.com/pradyunsg/vendoring which originated in https://github.com/pypa/pip/blob/19.0.3/tasks/vendoring/__init__.py (that code is much shorter).

It does something else entirely, but might have some good ideas.

**Comment by torsava** [comment]
Rebased on top of the version handling PR, all tests are passing.

**Comment by torsava** [comment]
Rebased on master after the version handling got merged.

**Comment by torsava** [comment]
There's again been a *lot* of activity, so I'l doing a rather time-consuming rebase.

**Comment by torsava** [comment]
I caught the exception and raised a new one from it, the output looks pretty good, @hroncok 


    Traceback (most recent call last):
      File "/home/torsava/git/upstream/rpm/scripts/pythondistdeps.py", line 140, in convert
        return OPERATORS[operator](name, operator, version_id)
      File "/home/torsava/git/upstream/rpm/scripts/pythondistdeps.py", line 115, in convert_ordered
        version_id = version_id[40]
    IndexError: string index out of range
    
    The above exception was the direct cause of the following exception:
    
    Traceback (most recent call last):
      File "/home/torsava/git/upstream/rpm/scripts/pythondistdeps.py", line 401, in <module>
        spec_list.append(convert(name, spec[0], spec[1]))
      File "/home/torsava/git/upstream/rpm/scripts/pythondistdeps.py", line 142, in convert
        raise RuntimeError("Cannot process Python package version: " + version_id) from exc
    RuntimeError: Cannot process Python package version `2.4.8.*` for name `python3.9dist(foobar47)`

**Comment by hroncok** [comment]
That's exactly how I would have done it. Keep rocking :P

**Comment by torsava** [comment]
@hroncok Ready for review!

**Comment by torsava** [comment]
I highly recommend reviewing commit-by-commit instead of by files changed.

File: .gitignore
Comment (hroncok) [inline .gitignore]: For consistency.

```suggestion
/tests/data/scripts_pythondistdeps/usr/
```

File: tests/README
Comment (hroncok) [inline tests/README]: What dependencies do i need? pytest, pyyaml and pip?

File: tests/test_scripts_pythondistdeps.py
Comment (hroncok) [inline tests/test_scripts_pythondistdeps.py]: *this* yaml file is what?

File: tests/data/scripts_pythondistdeps/test-requires.yaml
Comment (hroncok) [inline tests/data/scripts_pythondistdeps/test-requires.yaml]: Is it possible to have some tests for 3.10 as well to test double digits?

**Review by hroncok — COMMENTED** [review]
I couldn't resist and I've skimmed over the overall diff. The script diff is indeed unreadable due to indentation changes. Will review individual commit tmrw. Here are some nitpicks, feel free to ignore them before the more thorough review.

File: tests/test_scripts_pythondistdeps.py
Comment (hroncok) [inline tests/test_scripts_pythondistdeps.py]: returns a dict

File: .gitignore
Comment (torsava) [reply .gitignore]: Oh thank you, that would have bothered me if I realized it :D

File: tests/data/scripts_pythondistdeps/test-requires.yaml
Comment (torsava) [reply tests/data/scripts_pythondistdeps/test-requires.yaml]: Actually it is, for "sdist" I'm making the version up as I like. Ok, why not.

Comment (torsava) [reply tests/data/scripts_pythondistdeps/test-requires.yaml]: And I found a problem in the test-data getting script :laughing:

File: tests/test_scripts_pythondistdeps.py
Comment (torsava) [reply tests/test_scripts_pythondistdeps.py]: +1

File: tests/data/scripts_pythondistdeps/test-requires.yaml
Comment (torsava) [reply tests/data/scripts_pythondistdeps/test-requires.yaml]: Ugh, yaml takes `3.10` as a float and thus returns `3.1`. I guess I'll have to quote all the values.

File: tests/README
Comment (torsava) [reply tests/README]: Plus setuptools which is a dependency of pythondistdeps themselves, added.

**Comment by torsava** [comment]
Force pushed with fixes for all of those, thanks for noticing!

File: tests/test_scripts_pythondistdeps.py
Comment (torsava) [inline tests/test_scripts_pythondistdeps.py:150]: @hroncok I misread it when we were screen sharing. It's "SystemExit", i.e. it's fired when pip finishes successfully or not!

If we remove the except completely, the script ends after the one launch of pip.

And I already have an error message built in later, so doing a pass here even when pip fails seems to me ok. Output when there's a pip failure:

    ERROR: unknown command "installx" - maybe you meant "install"
    Problem occured while getting dist-info/egg-infofor package 'setuptools', version '41.6.0'for Python 2.7

Thus I think we should leave it as is.

Comment (hroncok) [reply tests/test_scripts_pythondistdeps.py:150]: Right. Thanks.

**Comment by torsava** [comment]
@hroncok Ready for review. Please take a look at all the fixup commits (some are added in the middle of the commit list).

And I've split the test-data into the relevant commits for easier readability.

Also, the trailing blank line in the commit is gone :D

File: scripts/pythondistdeps.py
Comment (hroncok) [inline scripts/pythondistdeps.py]:
```suggestion
            # dependency is relevant based on environment markers [1],
```

**Comment by hroncok** [comment]
> And I've split the test-data into the relevant commits for easier readability.

I see the major versions split, but not the normalized one.

Never mind. Please, do rebase the fixups, so I can do a final check. Thanks.

File: tests/data/scripts_pythondistdeps/test-requires.yaml
Comment (hroncok) [inline tests/data/scripts_pythondistdeps/test-requires.yaml]: tox 3.14 has:

```
install_requires =
    importlib-metadata >= 0.12, <1;python_version<"3.8"
```

I would like to see it tested: that importlib-metadata is required on 3.7 but not 3.8. Unfortunately 3.13.2 doesn't yet do this.

Or is there some other package here that does this?

**Comment by torsava** [comment]
Unrelated to this PR, but the version `11.0.0b2` is converted into `11~b2`, I gotta check if that's correct.

EDIT: according to [pyreq2rpm](https://github.com/gordonmessmer/pyreq2rpm/blob/master/tests/test_convert.py) it is correct:

        (['foobar', '==', '2.0.0b5'], 'foobar = 2~b5'),

And a cursory reading of PEP 440 looks like it confirms it, so all is well.

**Comment by torsava** [comment]
> I see the major versions split, but not the normalized one.
> 
> Never mind. Please, do rebase the fixups, so I can do a final check. Thanks.

They were in the fixup commit. Sorry, yes, there's too many commits :) I'll squash all and you'll see it more clearly.

File: scripts/pythondistdeps.py
Comment (torsava) [reply scripts/pythondistdeps.py]: Fixed manually

**Comment by hroncok** [comment]
> version `11.0.0b2` is converted into `11~b2`

yes, that is in fact correct.

The only problem is that `11~devX` sorts higher than `11~bX`. Hence an exception must be made to to convert to `11~~devX`. Luckily alpha and beta sort fine.

```
$ rpmdev-vercmp 11~b2 11~dev1
11~b2 < 11~dev1
$ rpmdev-vercmp 11~b2 11~0.dev1
11~b2 < 11~0.dev1
$ rpmdev-vercmp 11~b2 11~~dev1
11~b2 > 11~~dev1
```

Do you have a dev version package? See https://pypi.org/project/taskotron-python-versions/

File: tests/data/scripts_pythondistdeps/test-requires.yaml
Comment (torsava) [reply tests/data/scripts_pythondistdeps/test-requires.yaml]: I've switched to `tox 3.14`, and also `kubernetes` and `astroids` have these conditionals. I checked the outputs, they look good.

**Comment by torsava** [comment]
> Do you have a dev version package? See https://pypi.org/project/taskotron-python-versions/

What do you mean?

The package I found this in is `kubernetes==11.0.0b2`.

**Comment by torsava** [comment]
@hroncok All squashed and beautiful, take a look!

**Comment by hroncok** [comment]
>> Do you have a dev version package? See https://pypi.org/project/taskotron-python-versions/
>
> What do you mean?

Have a test with a `dev` versioned package as well. Not just `b`.

**Comment by hroncok** [comment]
Commits look good! All squashed and beautiful. Thanks.

**Comment by torsava** [comment]
> Have a test with a `dev` versioned package as well. Not just `b`.

Indeed I don't, and there's no such test on pyreq2rpm. Do you have a package that you know of? If not, I'll dig for some.

**Comment by hroncok** [comment]
Yes, see https://pypi.org/project/taskotron-python-versions/

**Comment by hroncok** [comment]
(Handle the dev test and fix in an extra commit please, it might change the behavior.)

**Comment by torsava** [comment]
Well, there's some deeper problem with developer versions :)

    $ find tests/data/scripts_pythondistdeps/usr/lib/python3.9/site-packages/taskotron-python-versions-0.1.dev6.dist-info | python scripts/pythondistdeps.py --provides
    python3.9dist(taskotron) = python

Wtf

**Comment by hroncok** [comment]
Wtf indeed.

**Comment by torsava** [comment]
Well, I'm getting it already as input from pkg_resources, so it goes even deeper.

**Comment by hroncok** [comment]
maybe taskotron-python-versions are broken?

**Comment by hroncok** [comment]
where is taskotron-python-versions-0.1.dev6.dist-info from?

**Comment by torsava** [comment]
> where is taskotron-python-versions-0.1.dev6.dist-info from?

Discussed off-github, we found that the test script was not saving the dist-info with the proper directory name, it should have been taskotron_python_versions_*. Fixed.

**Comment by torsava** [comment]
> The only problem is that `11~devX` sorts higher than `11~bX`. Hence an exception must be made to to convert to `11~~devX`. Luckily alpha and beta sort fine.

Done, PTAL @hroncok

**Comment by torsava** [comment]
We might also want to propose the `~~dev` handling to pyreq2rpm.

File: tests/test_scripts_pythondistdeps.py
Comment (hroncok) [inline tests/test_scripts_pythondistdeps.py]: English? :D

File: tests/data/scripts_pythondistdeps/pyreq2rpm.tests-2020.04.07.024dab0-py3.9.egg-info/requires.txt
Comment (hroncok) [inline tests/data/scripts_pythondistdeps/pyreq2rpm.tests-2020.04.07.024dab0-py3.9.egg-info/requires.txt:34]: Add dev versuon requirement somewhere here to also test requires, not only provides.

**Review by hroncok — CHANGES_REQUESTED** [review]

**Comment by hroncok** [comment]
The convert function is unusable because parse_version is not imported. We might want to reconsider the whole "make it importable" approach later and fix this. At least acknowledge in the commit message that this is a necessary step to make it reusable, but we are not there yet.

File: tests/test_scripts_pythondistdeps.py
Comment (torsava) [reply tests/test_scripts_pythondistdeps.py]: Omg, that was just my internal note from our review, not something I wanted to keep. Oops :D

File: tests/data/scripts_pythondistdeps/pyreq2rpm.tests-2020.04.07.024dab0-py3.9.egg-info/requires.txt
Comment (torsava) [reply tests/data/scripts_pythondistdeps/pyreq2rpm.tests-2020.04.07.024dab0-py3.9.egg-info/requires.txt:34]: +1

**Comment by torsava** [comment]
> The convert function is unusable because parse_version is not imported. We might want to reconsider the whole "make it importable" approach later and fix this. At least acknowledge in the commit message that this is a necessary step to make it reusable, but we are not there yet.

Why not just add `from pkg_resources import parse_version` inside `RpmVersion.__init__`? It's already imported at that point, so I don't think it's going to cause a slowdown.

**Comment by torsava** [comment]
> Why not just add `from pkg_resources import parse_version` inside `RpmVersion.__init__`? It's already imported at that point, so I don't think it's going to cause a slowdown.

Just ran it on the test data, and there's no slow down at all.

**Comment by hroncok** [comment]
> Why not just add from pkg_resources import parse_version inside RpmVersion.__init__? It's already imported at that point, so I don't think it's going to cause a slowdown.

Please don't. I'd tell you why but I only have like 5 seconds before another meeting.

I'll send a PR to https://github.com/gordonmessmer/pyreq2rpm/blob/master/pyreq2rpm/pyreq2rpm.py with changes I deem useful and they can be backported to the genarators. Don't hold this PR upstream on it.

**Comment by torsava** [comment]
> Please don't. I'd tell you why but I only have like 5 seconds before another meeting.

Ok, ping me when you have time. I still have to document it why it's not importable, so I need to understand it before opening an upstream PR.

>I'll send a PR to https://github.com/gordonmessmer/pyreq2rpm/blob/master/pyreq2rpm/pyreq2rpm.py with changes I > deem useful and they can be backported to the genarators. Don't hold this PR upstream on it.

I just tested that the `~~dev` works properly and it conforms to the [PEP ordering](https://www.python.org/dev/peps/pep-0440/#summary-of-permitted-suffixes-and-relative-ordering), so I was about to start on the PR. Do you want to make some other changes as well?

**Comment by hroncok** [comment]
> I was about to start on the PR

Please do, my changes will follow on top of that.

**Comment by torsava** [comment]
@hroncok Added some more tests. Ready for review.

**Comment by torsava** [comment]
Fixups squashed, tests passing. How about it, @hroncok ?

**Comment by torsava** [comment]
PR for pyreq2rpm: https://github.com/gordonmessmer/pyreq2rpm/pull/6

**Comment by hroncok** [comment]
Go upstream. Thank you!

**Review by hroncok — APPROVED** [review]
useless gif

**Comment by torsava** [comment]
@hroncok Thanks for all the fish!

**Comment by torsava** [comment]
Upstream PR: https://github.com/rpm-software-management/rpm/pull/1195

**Comment by hroncok** [comment]
Close this to avoid double notifications?

**Comment by torsava** [comment]
Closing this preparatory PR, the real PR is already open.

**Pull request closed without merging**
